usbus/msc: add initial Mass Storage Class support

dylad opened the pull request:
### Contribution description

This PR adds the initial support for Mass Storage Class in USBUS. This PR relies on the RIOT MTD implementation to implement the Mass Storage Class support. With the provided test application, a MTD device will be accessible as a normal storage device on your host computer.
Read and Write operations are allowed.
Multiple LUNs are supported so several MTD devices can be exported through USB.
The MSC relies on SCSI protocol to operate.

Currently there are some limitations:
    Supported host : Linux & Windows (macOS is untested)
    MSC cannot be used if MTD page size > 4096
    MTD device must have at least 512 bytes of memory to be exported.

Please be aware that performance are not so great.

### Testing procedure
Flash `tests/usbus_msc` application on a board with at least one MTD device.
Once the shell has started, prepare one or several MTD devices to be exported using `add_lun` command.
Once ready, start the USB connection with `usb_attach`

All MTD exported should appear as` /dev/sdX` on Linux.

### Issues/PRs references
Supersede #15941

benpicco commented on sys/include/usb/usbus/msc/scsi.h:
```diff
@@ -0,0 +1,293 @@
+ *
+ * @see Read10 or Write10 Command from SCSI Primary Command
+ *
+ * @warning uint16_t and uint32_t will be BIG ENDIAN
```
better use `network_uint16_t`/`network_uint32_t` then

dylad replied:
Good idea !
But the `network_*` label puzzle me a bit.
How about `be_uint16_t` and so instead ?

benpicco replied:
> How about `be_uint16_t` and so instead?

uh I didn't know we had both - yea `network_uint16_t` is just a typedef of `be_uint16_t` anyway, I just had never seen the latter.

dylad replied:
I've switched to `be_uint*_t` where applicable. I'll run some tests later this week to ensure I didn't break something.

benpicco commented on sys/usb/usbus/msc/msc.c:
```diff
@@ -0,0 +1,437 @@
+        if (mtd->page_size <= block_size) {
+            /* Write one or multiple pages */
+            printf("WRITE:%d bytes\n", block_size);
+            ret = mtd_write_page(mtd, msc->buffer, msc->block, 0,
```
Since you are writing a full page, you can use `mtd_write_page_raw()` and avoid the read-modify-write cycle

benpicco replied:
Ah I see you actually want this for cases where the block size is > 4096.
We should probably add a fast-path to `mtd_write_page()` then to skip the read-modify-write cycle if the whole sector is written - #19252

dylad replied:
I wanted to implement both `mtd_write_page()` and `mtd_write_page_raw()` for some time but since this PR stalled I think we should just stick to one of them for now.
But I don't mind to switch to `mtd_write_page_raw()` if you think it is more suitable.

benpicco commented on sys/usb/usbus/msc/msc.c:
```diff
@@ -0,0 +1,437 @@
+    int ret;
+
+    /* Check if we have a block to read and transfer */
+    //printf("block_nb:%d\n", msc->block_nb);
```
either `DEBUG()` or remove it

benpicco commented on sys/usb/usbus/msc/msc.c:
```diff
@@ -0,0 +1,437 @@
+
+/* Internal buffer to handle size difference between MTD layer and USB
+   endpoint size as some MTD implementation (like mtd_sdcard) doesn't allow
+   endpoint size transfer */
```
Huh the MTD layer should take care of splitting up the transfers.
What do you mean here?

dylad replied:
This buffer is used to store data coming from the endpoint. Since USB endpoints transfers (for Full-Speed device) are capped at 64 bytes, this buffer stores each transfer until it reaches the MTD device flash page size.
For example, if the MTD page size is 512 bytes. We need to have 8 endpoints transfer of 64 bytes to fill this buffer so we can trigger a write to the MTD device.
Moreover, since I cannot report the real page size of the MTD device if it's less than 512 bytes (e.g for EEPROM). This buffer will then have to store multiples pages at once.
But I agree that the wording is weird. I can come up with something better.

benpicco replied:
You don't *need* to do this if you erase the sector on the first block and always write a whole sector. (`mtd_write_page_raw()` will not to the read-modify-write but just write and you don't need to write the whole sector at once).

There might be a slight performance benefit to it, but I'm not sure.

If you are the sole user of the MTD device you could also use `mtd->work_area` which is a per-device, dynamically allocated sector-sized buffer.

That would be a bit hacky though - but I also don't see a problem with just writing 64 bytes at a time.

dylad replied:
> You don't need to do this if you erase the sector on the first block and always write a whole sector. 

Unfortunately, I am not sure I can guarantee I'll always write a whole sector. USB Host will use the reported page size and doesn't know about the sector size of the MTD device. So it may write only one or a few pages.

benpicco replied:
Then how do you decide when to flush?
Or are those writes just lost if the sector is not completed?

dylad replied:
> Or are those writes just lost if the sector is not completed?

No, USB Host expects the USB device to write the data that were send. So if the Host only write a page, we must do a read-modify-write operation on our side.

benpicco replied:
Ah now I get it!
USB will always transfer a page¹ (in 64 byte chunks) but since we can only erase a sector, we have to do a read-modify write cycle.

Buffering the page brings that down to one.

[1] a 'USB page' that will usually be just as big or bigger then the sector size. In that case we don't need the RMW. Only if the MTD page is greater than 4096 bytes we have chose a smaller USB page and resort to RMW.

I would propose the following:

 - if the 'USB page' is smaller than the sector size, allocate a temporary buffer as you do here, but you may use `malloc` (as that's already used by `mtd_pagewise` - I think `malloc` on init is fine.
 - if the 'USB page' is greater or equal the sector size, don't allocate a buffer and write the bytes as you get them with `mtd_write_page_raw()`, ereasing the sector at the first block.

That way you don't have the memory overhead in the common case and don't have to tune config variables to fit the storage.

What do you think?

dylad replied:
> USB will always transfer a page¹ (in 64 byte chunks) but since we can only erase a sector, we have to do a read-modify write cycle.

Very true.

> Buffering the page brings that down to one RMW cycle.

That's the idea.

To sum up, there are 3 cases to properly handle:

- USB reported page size == MTD page size

  In this case, (512, 1024, 2048, 4096), this is pretty easy. We store the 64 bytes chunks into a buffer (allocated dynamically or not) Since we have a whole MTD page (not a MTD sector) we can write it directly to the memory. (And we rely on the MTD implementation to handle the erase part).
- USB reported page size > MTD page size

This is the case for a few MTD type like EEPROM. In this case, we report an USB page size of 512 (the minimum allowed). So if MTD page size is 256. We have to write two MTD pages for each USB page. 
This is also why we cannot export MTD device which have less than 512 bytes of memory. (Well we could cheat but this will just add more complexity for a small corner case and I believe this PR is already complex enough...)

- USB reported page size < MTD page size

This is a more tricky one, which is not yet support by this PR (even though there is some code laying around). For this case, we need to do some RMW cycle. But this will belong to a followup PR.

dylad replied:
> I would propose the following:
    if the 'USB page' is smaller than the sector size, allocate a temporary buffer as you do here, but you may use malloc (as that's already used by mtd_pagewise - I think malloc on init is fine.
    if the 'USB page' is greater or equal the sector size, don't allocate a buffer and write the bytes as you get them with mtd_write_page_raw(), ereasing the sector at the first block.

> That way you don't have the memory overhead in the common case and don't have to tune config variables to fit the storage.

> What do you think?

Performance-wise, the second idea is great but I need to check if it won't make the code unreadable because of the corner cases.
For the first idea, I think this will the most common pattern. I can switch to dynamically allocated buffer since we let the user chose which MTD to export through the test application. malloc() can then be call only and only if the user request the MTD device to be exported. This may save some memory on boards which have a several MTD that one can export but only need one to be exported.

benpicco replied:
But why tie this to page size? A page is trust the largest transfer unit, but the MTD will already take care of that, so on the application level it's rather arbitrary.

Sector size is the much more interesting value as it's the erase unit. If you report the sector size (and you will always get a full sector through MSC) you can just erase the sector at the start of the transfer and then write the 64 byte blocks as they come (without the RMW).

dylad replied:
I see your point. This will increase performance for some MTD devices. So you suggest to erase the sector from USBUS then call `mtd_write_page_raw()` for each USB endpoints chunks received to avoid storing the content of a page ?
I can try to give it a go. I don't think such rework would be too painful.
Obviously, this won't work at all for sector size > 4096.

dylad replied:
I am making progress here, I have a working version using sector size instead of page size.
I am now looking at the buffer side and I think we shouldn't drop it even if the USB block == MTD sector size:
This might impact performance on read operation. Currently, a single `mtd_read()` is perform and store in the buffer. Then we transfer the buffer through USB chunks. Dropping the buffer would force us to call `mtd_read()` for each USB transfer chunks and it seems sub-optimal to me.

Regarding the dynamical allocation of the buffer, I am not against. This can be done during the `USB_MSC_SETUP_REQ_GML` request. This way we can iterate through all exported MTD devices to USB. Determine which device has the greatest sector size and allocate this sector size to the buffer. No need for user configuration.

Note that even if multiple LUNs are used at the same time. Operation can only be done one LUN at a time. So having one buffer is completely fine for all LUNs (read or write operations).

benpicco reviewed (commented):
I don't know much about the USB part, but I can take a look at the MTD part.

benpicco commented on sys/usb/usbus/msc/scsi.c:
```diff
@@ -0,0 +1,386 @@
+    /* Convert big endian data from SCSI to little endian */
+    /* Get first block number to read from */
+    if (mtd->page_size < 512) {
+        msc->block = byteorder_swapl(pkt->blk_addr.u32) * (512 / mtd->page_size);
```
```suggestion
        msc->block = byteorder_ntohl(pkt->blk_addr) * (512 / mtd->page_size);
```

benpicco commented on sys/usb/usbus/msc/scsi.c:
```diff
@@ -0,0 +1,386 @@
+        assert(0);
+    }
+    else {
+        msc->block = byteorder_swapl(pkt->blk_addr.u32);
```
```suggestion
        msc->block = byteorder_ntohl(pkt->blk_addr);
```

benpicco commented on sys/usb/usbus/msc/scsi.c:
```diff
@@ -0,0 +1,386 @@
+         * apply some logic and let host believe this is a 512
+         * page size device and let RIOT write several page in
+         * a row to reach 512 bytes so everyone is happy */
+        msc->block = byteorder_swapl(pkt->blk_addr.u32) * (512 / mtd->page_size);
```
```suggestion
        msc->block = byteorder_ntohl(pkt->blk_addr) * (512 / mtd->page_size);
```

benpicco commented on sys/usb/usbus/msc/scsi.c:
```diff
@@ -0,0 +1,386 @@
+
+    /* Only support one list, size of the whole struct minus the 4 first bytes */
+    pkt->list_length = len - 4;
+    pkt->blk_nb = byteorder_ltobl((le_uint32_t)blk_nb);
```
```suggestion
    pkt->blk_nb = byteorder_htons(blk_nb);
```

dylad replied:
It hurts the brain.

dylad commented:
Lastest commit changes the following:
- Use MTD sector size instead of MTD page size as USB block.
- Dynamically allocate the internal buffer depending on the MTD with the greatest sector size value.

benpicco commented on sys/include/usb/usbus/msc.h:
```diff
@@ -0,0 +1,119 @@
+int usbus_msc_init(usbus_t *usbus, usbus_msc_device_t *handler);
+
+/**
+ * @brief Register a MTD device as a MSC LUN
```
Would be good to have LUN explained at least once

dylad replied:
Done here.

benpicco commented on sys/usb/usbus/msc/msc.c:
```diff
@@ -0,0 +1,453 @@
+
+    switch (setup->request) {
+    case USB_MSC_SETUP_REQ_GML:
+        uint8_t data = _get_lun(msc);
```
```suggestion
        data = _get_lun(msc);
```

benpicco commented on sys/usb/usbus/msc/msc.c:
```diff
@@ -0,0 +1,453 @@
+    usbus_msc_device_t *msc = container_of(handler, usbus_msc_device_t,
+                                           handler_ctrl);
+
+    switch (setup->request) {
```
```suggestion
    uint8_t data;
    switch (setup->request) {
```

benpicco commented on sys/include/usb/usbus/msc.h:
```diff
@@ -0,0 +1,119 @@
+#define USBUS_MSC_EXPORTED_DEV MTD_NUMOF
+#endif /* USBUS_MSC_EXPORTED_DEV */
+
+typedef enum {
+    WAITING,
+    WAIT_FOR_TRANSFER,
+    DATA_TRANSFER_IN,
+    DATA_TRANSFER_OUT,
+    DATA_TRANSFER_LAST,
+    GEN_CSW
+} usbus_msc_state_t;
```
Eh, doxygen is right that those should be documented.
Someone (or you yourself) touching this code a few years down the line will be thankful to have an explanation what those states mean.

dylad replied:
Added comments and remove `DATA_TRANSFER_LAST` which was unused. Probably some leftover from long ago.

benpicco commented on sys/include/usb/usbus/msc.h:
```diff
@@ -0,0 +1,119 @@
+#endif
+
+#ifndef USBUS_MSC_EXPORTED_DEV
+#define USBUS_MSC_EXPORTED_DEV MTD_NUMOF
```
nitpick, but that sounds like you would say which device to export here.
How about
```suggestion
#define USBUS_MSC_EXPORTED_NUMOF MTD_NUMOF
```

dylad replied:
Updated.

benpicco commented on sys/usb/usbus/msc/msc.c:
```diff
@@ -0,0 +1,453 @@
+    }
+
+    /* No more slots available */
+    return -EAGAIN;
```
```suggestion
    return -ENFILE;
```
might be a better fit

benpicco reviewed (commented):
Nice one, works like a charm on my `same54-xpro` :smiley: 
Read speeds are ~600kB/s for both SD card and SPI flash, I wonder if this could benefit from #17091 as a follow-up.

Some doc comments mostly:

dylad commented:
I think this PR would benefit #17091. At least, it will lower a bit the code complexity of this PR.
But since #17091 doesn't have any use case in master, I guess it's better to have this one first so @bergzand could give it a quick rebase right after ;)

benpicco commented on sys/usb/usbus/msc/msc.c:
```diff
@@ -0,0 +1,454 @@
+    }
+
+    /* No more slots available */
+    free(msc->buffer);
```
```suggestion
    free(msc->buffer);
    msc->buffer = NULL;
```

benpicco commented on sys/usb/usbus/msc/scsi.c:
```diff
@@ -0,0 +1,386 @@
+    pkt->removable = 1;
+    pkt->version = 0x01;
+    pkt->length = len - 4;
+    pkt->flags[0] = 0x80;
```
What flag is `0x80`?

dylad replied:
This flag is SCCS (SCC Supported) bit. IIRC, this means Storage Controller Component.
This bit is commonly set but I forgot to create a `define` for this one

dylad replied:
Was not needed in the end, so It got removed.

benpicco commented on sys/usb/usbus/msc/scsi.c:
```diff
@@ -0,0 +1,386 @@
+    pkt->type = 0;
+    pkt->removable = 1;
+    pkt->version = 0x01;
+    pkt->length = len - 4;
```
Why `- 4`?

dylad replied:
This is the additional packet length minus the "default" data of this packet.
a SCSI inquiry response will always include the 4 first bytes, but you may have to pass additional information. Thus, you pass the number additional bytes through this field.

I'll add a comment.

benpicco commented on sys/usb/usbus/msc/scsi.c:
```diff
@@ -0,0 +1,386 @@
+    /* Report the size of the struct minus the 7 first bytes */
+    pkt->add_len = len - 7;
+    pkt->asc = 0x30;
+    pkt->ascq = 0x01;
```
What are those magic numbers?

dylad replied:
Is undocumented black magic numbers, a valid answer here ?

benpicco replied:
Where did you get them from?

dylad replied:
Hmm, I think it was on some source code provided by ST for a STM32F1 that I used back in 2019.
I did have a look at Zephyr a few months ago too, but it was the same. (Just hexadecimal value used to fill their buffer).
I'll have a look at it tonight. I have a bunch of USB MSC related document on my home laptop. I'll see if I can give you more insight then.

benpicco reviewed (commented):
Please squash (and rebase)

btw, https://github.com/RIOT-OS/RIOT/pull/17091 implements URBs for CDC ECM.

dylad commented:
Rebased and squashed.
@benpicco Regarding the ACS, ACSQ "black magic" stuff: I've just add a comment for now.
These values are in fact used to report to the host additional information regarding different kind of failures.
I'll make use of this in a later PR, let's stick to default values for now, as more work is required.

benpicco commented on tests/usbus_msc/Makefile:
```diff
@@ -0,0 +1,29 @@
+
```
This will also need a `Makefile.ci`
You can create it with

    make generate-Makefile.ci

benpicco commented:
bors merge

benpicco commented:
bors cancel
bors merge

benpicco commented:
bors try

dylad commented:
Do we need a bors stop drinking again ?

benpicco commented:
I think bors already went to sleep

bors ping

dylad commented:
@benpicco Thanks for the review & merge !

benpicco commented on tests/usbus_msc/main.c:
```diff
@@ -0,0 +1,181 @@
+    /* Initialize Mass Storage Class */
+    usbus_msc_init(&usbus, &msc);
+    /* Create USBUS thread */
+    usbus_create(_stack, USBUS_STACKSIZE, USBUS_PRIO, USBUS_TNAME, &usbus);
```
Why is this needed here if we already have `auto_init_usbus`?

benpicco commented on tests/usbus_msc/main.c:
```diff
@@ -0,0 +1,181 @@
+    assert(usbdev);
+
+    /* Initialize basic usbus struct, don't start the thread yet */
+    usbus_init(&usbus, usbdev);
```
Same here

benpicco commented on tests/usbus_msc/main.c:
```diff
@@ -0,0 +1,181 @@
+    usbus_init(&usbus, usbdev);
+
+    /* Initialize Mass Storage Class */
+    usbus_msc_init(&usbus, &msc);
```
Should this be moved to `auto_init_usb.c`?

benpicco reviewed (commented):
I've been trying to combine this with `stdio_cdc_acm` but so far no luck.

Is there some inherent conflict or is the issue just that we initialize the USBus twice?

dylad commented:
Because It was easier to setup things this way.
The idea was to merge this stuff and let the users decided which MTD devices they want to use at runtime. And yes, it doesn't get along with CDC for now because I still need a bit of time to find the right solution for this.
Nevertheless, I should have a describe it in the Documentation, my bad.

I don't like the idea of initialize and export all MTD devices unconditionally. And I don't like the idea of an always ON MSC implementation either. It'll just increase the overall USB transfer for nothing with "ping" transfer aka TEST_UNIT_READY.

To me, the MSC and the MTD devices selection should be a user runtime choice.
Obviously, there are some uses cases to export all available MTD devices (this is why I implemented the multiple LUNs support) and to have a MSC always ON (hobbyst or academic playground). But I don't think it should be the default behavior.
